### PR TITLE
[Website] Euphoria roadmap.

### PR DESCRIPTION
**Pull request by dmvk**
First shot on euphoria roadmap.
------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

It will help us expedite review of your Pull Request if you tag someone (e.g. `@username`) to look at it.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_GradleBuild/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_GradleBuild/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Gradle/lastCompletedBuild/) [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark_Gradle/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | --- | ---

### CODE REVIEW

File: website/src/roadmap/euphoria.md [hunk 1]
```diff
@@ -23,7 +23,74 @@ limitations under the License.
...
+```
+
+After:
+
+```java
+final PCollection<KV<String, Integer>> counts = ...;
+
+SumByKey.named("SummarizeCounts")
+  .ofKeyed(counts)
+  .valueBy(KV::getValue)
```
Comment (je-ik) [inline website/src/roadmap/euphoria.md]: I suppose this could be also omitted (or made default)?

File: website/src/roadmap/euphoria.md [hunk 2]
```diff
@@ -23,7 +23,74 @@ limitations under the License.
 Easy to use Java 8 DSL for the Beam Java SDK. Provides a high-level abstraction of Beam transformations, which is both easy to read and write. Can be used as a complement to existing Beam pipelines (convertible back and forth). You can have a glimpse of the API at [WordCount example]({{ site.baseurl
 }}/documentation/sdks/java/euphoria/#wordcount-example).
 
-This roadmap is in progress. In the meantime, here are some resources:
-
 - JIRA: [dsl-euphoria](https://issues.apache.org/jira/browse/BEAM-4366?jql=project%20%3D%20BEAM%20AND%20component%20%3D%20dsl-euphoria) / [BEAM-3900](https://issues.apache.org/jira/browse/BEAM-3900)
-- Contact: [David Moravek](mailto:david.moravek@gmail.com)
\ No newline at end of file
+- Contact: [David Moravek](mailto:dmvk@apache.org)
```
Comment (VaclavPlajt) [inline website/src/roadmap/euphoria.md]: Can we have documentation link here? Currently we can get there through `WordCount example`, but that is not really straightforward.

File: website/src/roadmap/euphoria.md (hunk 1 above)
Comment (VaclavPlajt) [reply website/src/roadmap/euphoria.md]: We cannot omit that I'm afraid. That would change the API. And it seems to me that it is useful to have key and value extractors in API. One can skip one `MapElements` operation.

Comment (je-ik) [reply website/src/roadmap/euphoria.md]: I'm not sure I understand. If we have keyed `PCollection`, why do we make default only the `key`, and not the `value` from the `KeyValue`?

Comment (VaclavPlajt) [reply website/src/roadmap/euphoria.md]: Sorry, my bad. I misunderstand you comment.

**Comment by mareksimunek** [comment]
LGTM

File: website/src/roadmap/euphoria.md (hunk 1 above)
Comment (dmvk) [reply website/src/roadmap/euphoria.md]: I guess this make sense, we'd still need optional `valueBy` method though. 👍

File: website/src/roadmap/euphoria.md (hunk 2 above)
Comment (dmvk) [reply website/src/roadmap/euphoria.md]: 👍 makes sense

**Comment by kennknowles** [comment]
Might tag @melap here too.

**Comment by dmvk** [comment]
@kennknowles thanks for suggestions, I've simplified the roadmap. We'll bring up the pipeline sampling on dev list in following weeks ;)

@je-ik Complete join sematics sounds like a great feature, but this is really low on our priority list right now. If you would like to work on this, we can put it in the roadmap :) I'll turn your other suggestions into jira issues as they mostly seem like small changes.

**Comment by je-ik** [comment]
@dmvk I'm definitely interested in working on that in long term. I'd put it into the roadmap (without too much details, because there are lot of things that have to be figured out) as something like "table-stream joins". The roadmap doesn't have any term associated with individual items, right? I see it only as a statement of what directions we want to move and in this sense the table-stream join is in my opinion a good candidate for that.

**Comment by dmvk** [comment]
Join semantics added to roadmap in https://github.com/apache/beam/pull/7100

**Pull request merged**